In [3]:
import torch

if torch.cuda.is_available():
    device = torch.device("cuda")
    print(f"GPU model: {torch.cuda.get_device_name(0)}")
    print(f"GPU device count: {torch.cuda.device_count()}")
else:
    print("No GPU available.")
  

GPU model: Tesla T4
GPU device count: 1


# Data Pre-processing

First we will begin by pre-processing the data. In this case the sample data comes with the following features/columns: 

|   **Column**   |   **Desciption**   |
|:-:	         |:--	              |
|   STUD	|   Study Number 	|
|   DSFQ	|   Dosing Frequency	|
|   PTNM	|   Patient Number	|
|   CYCL	|    Dosing Cycles	|
|   AMT	|   Dosing Amounts	|
|   TIME	|   Time in Hours Since the Experiment Began for one Individual	|
|   TFDS	|   Time in Hours Since the Last Dosing|
|   DV/PK_timeCourse	|   The Observations of PK  	|

Let us first import the data into the notebook and observe these features of the dataset

In [4]:
from google.colab import files
 
 
uploaded = files.upload()

Saving sim_data.txt to sim_data.txt


In [5]:
# Importing required libraries for data pre-processing

import pandas as pd
import numpy as np
import io


#Reading the csv file with the data

data_complete = pd.read_csv(io.BytesIO(uploaded['sim_data.txt'])) #change to sim_data.csv 
data_complete


# data_complete = pd.read_csv("/Users/rishabhgoel/Desktop/NeuralODE_Paper_Supplementary_Code/ExampleData/sim_data.csv", na_values='.')

# data_complete

STUD   PTNM  DSFQ  CYCL    AMT    TIME    TFDS         DV
0     1000.0    1.0   3.0   1.0  296.6     0.0     0.0  20.382000
1     1000.0    1.0   3.0   1.0    0.0    24.0    24.0  73.148000
2     1000.0    1.0   3.0   1.0    0.0   216.0   216.0  19.764000
3     1000.0    1.0   3.0   1.0    0.0   504.0   504.0   3.219900
4     1000.0    1.0   3.0   2.0  288.0   504.0     0.0   3.219900
...      ...    ...   ...   ...    ...     ...     ...        ...
5371  3000.0  200.0   3.0  15.0    0.0  7416.0   360.0   6.091200
5372  3000.0  200.0   3.0  15.0    0.0  7584.0  7584.0   2.043100
5373  3000.0  200.0   3.0  17.0  259.2  8064.0     0.0   0.090115
5374  3000.0  200.0   3.0  17.0    0.0  8088.0    24.0  53.990000
5375  3000.0  200.0   3.0  17.0    0.0  8256.0   192.0  17.255000

[5376 rows x 8 columns]

There are 5376 observations and 8 features that we are looking at. Now let us begin the pre-processing the data. 

Our sample data only has the relevant columns for the model. However, in the data we receive from patients there will be many more features so it is important for us to select the correct features. Thus, let us begin by creating a place holder for the features that are important to the model.

In [6]:
#variable for colummns that we will eventually select from the raw dataset we receive

select_cols = ["STUD", "DSFQ", "PTNM", "CYCL", "AMT", "TIME", "TFDS", "DV"]

#Selecting the relevant columns from the dataframe

data_complete = data_complete[select_cols]

We then start filtering data based on multiple parameters:
    
1. Dosing Cycle < 100

In [8]:
# filtering for only the rows in the dataframe with a dosing cycle less than 100

data_complete = data_complete[data_complete.CYCL < 100]
data_complete

STUD  DSFQ   PTNM  CYCL    AMT    TIME    TFDS         DV
0     1000.0   3.0    1.0   1.0  296.6     0.0     0.0  20.382000
1     1000.0   3.0    1.0   1.0    0.0    24.0    24.0  73.148000
2     1000.0   3.0    1.0   1.0    0.0   216.0   216.0  19.764000
3     1000.0   3.0    1.0   1.0    0.0   504.0   504.0   3.219900
4     1000.0   3.0    1.0   2.0  288.0   504.0     0.0   3.219900
...      ...   ...    ...   ...    ...     ...     ...        ...
5371  3000.0   3.0  200.0  15.0    0.0  7416.0   360.0   6.091200
5372  3000.0   3.0  200.0  15.0    0.0  7584.0  7584.0   2.043100
5373  3000.0   3.0  200.0  17.0  259.2  8064.0     0.0   0.090115
5374  3000.0   3.0  200.0  17.0    0.0  8088.0    24.0  53.990000
5375  3000.0   3.0  200.0  17.0    0.0  8256.0   192.0  17.255000

[5376 rows x 8 columns]

We then convert the "PTNM" column to an integer type using the astype() method, and then using the map() method to apply a format string to each value in the column. The format string "{:05d}" specifies that each value should be formatted as a zero-padded integer with a width of 5 digits.

For example, if the "PTNM" column originally contained the values [1, 10, 100, 1000], after this line of code is executed, the "PTNM" column would contain the values ['00001', '00010', '00100', '01000'].

This line of code is useful for standardizing the format of the values in the "PTNM" column, which can make it easier to perform operations on the column and to compare values within the column. It can also be useful for preparing the data for downstream model applications that require the data to be in a particular format.

In [9]:
# formatting Patient Number for Unique Identifier
data_complete["PTNM"] = data_complete["PTNM"].astype("int").map("{:05d}".format)
data_complete

STUD  DSFQ   PTNM  CYCL    AMT    TIME    TFDS         DV
0     1000.0   3.0  00001   1.0  296.6     0.0     0.0  20.382000
1     1000.0   3.0  00001   1.0    0.0    24.0    24.0  73.148000
2     1000.0   3.0  00001   1.0    0.0   216.0   216.0  19.764000
3     1000.0   3.0  00001   1.0    0.0   504.0   504.0   3.219900
4     1000.0   3.0  00001   2.0  288.0   504.0     0.0   3.219900
...      ...   ...    ...   ...    ...     ...     ...        ...
5371  3000.0   3.0  00200  15.0    0.0  7416.0   360.0   6.091200
5372  3000.0   3.0  00200  15.0    0.0  7584.0  7584.0   2.043100
5373  3000.0   3.0  00200  17.0  259.2  8064.0     0.0   0.090115
5374  3000.0   3.0  00200  17.0    0.0  8088.0    24.0  53.990000
5375  3000.0   3.0  00200  17.0    0.0  8256.0   192.0  17.255000

[5376 rows x 8 columns]

Now we create a new column called "ID" in the pandas DataFrame data_complete. The "ID" column is being created by concatenating two existing columns in ```data_complete```: "STUD" and "PTNM".

First, the "STUD" column is being converted to an integer data type using the ```astype()``` method with the argument "int". Then, the "STUD" column is being converted to a string data type using the ```astype()``` method with the argument "str". This ensures that the values in the "STUD" column are in string format.

Next, the values in the "STUD" and "PTNM" columns are being concatenated using the + operator, which joins the two strings together. This creates a new string for each row in the DataFrame, which is assigned to the "ID" column.

For example, if the "STUD" column contained the values ['100', '101', '102', '103'] and the "PTNM" column contained the values ['00001', '00010', '00100', '01000'], then after this line of code is executed, the "ID" column would contain the values ['10000001', '10100010', '10200100', '10301000'].

This line of code is useful for creating a unique identifier for each row in the DataFrame based on the values in the "STUD" and "PTNM" columns. This can be useful for identifying and tracking individual records, as well as for linking data across multiple datasets.

In [10]:
# Creating a unique identifier column, 'ID', and formatting it based on Patient Number and Study Number
data_complete["ID"] = data_complete["STUD"].astype("int").astype("str") + data_complete["PTNM"]
data_complete

STUD  DSFQ   PTNM  CYCL    AMT    TIME    TFDS         DV         ID
0     1000.0   3.0  00001   1.0  296.6     0.0     0.0  20.382000  100000001
1     1000.0   3.0  00001   1.0    0.0    24.0    24.0  73.148000  100000001
2     1000.0   3.0  00001   1.0    0.0   216.0   216.0  19.764000  100000001
3     1000.0   3.0  00001   1.0    0.0   504.0   504.0   3.219900  100000001
4     1000.0   3.0  00001   2.0  288.0   504.0     0.0   3.219900  100000001
...      ...   ...    ...   ...    ...     ...     ...        ...        ...
5371  3000.0   3.0  00200  15.0    0.0  7416.0   360.0   6.091200  300000200
5372  3000.0   3.0  00200  15.0    0.0  7584.0  7584.0   2.043100  300000200
5373  3000.0   3.0  00200  17.0  259.2  8064.0     0.0   0.090115  300000200
5374  3000.0   3.0  00200  17.0    0.0  8088.0    24.0  53.990000  300000200
5375  3000.0   3.0  00200  17.0    0.0  8256.0   192.0  17.255000  300000200

[5376 rows x 9 columns]

In [11]:
# Creating a dataframe with the maximum time for each patient in each study (Essentially presenting the time of the last observed dose)
time_summary = data_complete[["ID", "TIME"]].groupby("ID").max().reset_index()

# Creating a new dataframe with only the IDs for patients with a time of the last observed dose greater than 0 (eliminating errors/outliers)
selected_ptnms = time_summary[time_summary.TIME > 0].ID

# Only selecting IDs with the last observed dose greater than 0 in our main table called data_complete
data_complete = data_complete[data_complete.ID.isin(selected_ptnms)]
data_complete

STUD  DSFQ   PTNM  CYCL    AMT    TIME    TFDS         DV         ID
0     1000.0   3.0  00001   1.0  296.6     0.0     0.0  20.382000  100000001
1     1000.0   3.0  00001   1.0    0.0    24.0    24.0  73.148000  100000001
2     1000.0   3.0  00001   1.0    0.0   216.0   216.0  19.764000  100000001
3     1000.0   3.0  00001   1.0    0.0   504.0   504.0   3.219900  100000001
4     1000.0   3.0  00001   2.0  288.0   504.0     0.0   3.219900  100000001
...      ...   ...    ...   ...    ...     ...     ...        ...        ...
5371  3000.0   3.0  00200  15.0    0.0  7416.0   360.0   6.091200  300000200
5372  3000.0   3.0  00200  15.0    0.0  7584.0  7584.0   2.043100  300000200
5373  3000.0   3.0  00200  17.0  259.2  8064.0     0.0   0.090115  300000200
5374  3000.0   3.0  00200  17.0    0.0  8088.0    24.0  53.990000  300000200
5375  3000.0   3.0  00200  17.0    0.0  8256.0   192.0  17.255000  300000200

[5376 rows x 9 columns]

In [12]:
# filling in missing values in the "AMT" column with Dosing Amounts equal to 0
data_complete["AMT"] = data_complete["AMT"].fillna(0)


#Renaming the column DV to PK_timeCourse so that the name is more self-explanatory
data_complete = data_complete.rename(columns={"DV": "PK_timeCourse"})


# Duplicating PK_timeCourse column to make changes to it separately
data_complete["PK_round1"] = data_complete["PK_timeCourse"]

# Changing the PK_round1 such that if the dosing frequency is once weekly and the time of the last 
# observed dose is greater than 168hrs the PK for round 1 is zero. Similarly, if the dosing frequency 
# is once weekly and the time of the last observed dose is greater than 504hrs the PK for round 1 is zero
data_complete.loc[(data_complete.DSFQ == 1) & (data_complete.TIME >= 168), "PK_round1"] = 0
data_complete.loc[(data_complete.DSFQ == 3) & (data_complete.TIME >= 504), "PK_round1"] = 0

# filling in missing values in the "PK_round1" column with 0
data_complete["PK_round1"] = data_complete["PK_round1"].fillna(0)

# filling in missing values in the "PK_timeCourse" column with -1
data_complete["PK_timeCourse"] = data_complete["PK_timeCourse"].fillna(-1)
data_complete

STUD  DSFQ   PTNM  CYCL    AMT    TIME    TFDS  PK_timeCourse  \
0     1000.0   3.0  00001   1.0  296.6     0.0     0.0      20.382000   
1     1000.0   3.0  00001   1.0    0.0    24.0    24.0      73.148000   
2     1000.0   3.0  00001   1.0    0.0   216.0   216.0      19.764000   
3     1000.0   3.0  00001   1.0    0.0   504.0   504.0       3.219900   
4     1000.0   3.0  00001   2.0  288.0   504.0     0.0       3.219900   
...      ...   ...    ...   ...    ...     ...     ...            ...   
5371  3000.0   3.0  00200  15.0    0.0  7416.0   360.0       6.091200   
5372  3000.0   3.0  00200  15.0    0.0  7584.0  7584.0       2.043100   
5373  3000.0   3.0  00200  17.0  259.2  8064.0     0.0       0.090115   
5374  3000.0   3.0  00200  17.0    0.0  8088.0    24.0      53.990000   
5375  3000.0   3.0  00200  17.0    0.0  8256.0   192.0      17.255000   

             ID  PK_round1  
0     100000001     20.382  
1     100000001     73.148  
2     100000001     19.764  
3     100000001      0.000  
4     100000001      0.000  
...         ...        ...  
5371  300000200      0.000  
5372  300000200      0.000  
5373  300000200      0.000  
5374  300000200      0.000  
5375  300000200      0.000  

[5376 rows x 10 columns]

In [13]:
# Removing rows where the "AMT" column is 0 and the "TIME" column is also 0.
data_complete = data_complete[~((data_complete.AMT == 0) & (data_complete.TIME == 0))]

# Keeping the last row for all patients with duplicate values for time in hours 
data_complete.loc[data_complete[["PTNM", "TIME"]].duplicated(keep="last"), "AMT"] = \
    data_complete.loc[data_complete[["PTNM", "TIME"]].duplicated(keep="first"), "AMT"].values

# Keeping the first row for all observations with duplicate values for all features apart from patient and time in hours 
data_complete = data_complete[~data_complete[["PTNM", "TIME"]].duplicated(keep="first")]

data_complete

STUD  DSFQ   PTNM  CYCL    AMT    TIME    TFDS  PK_timeCourse  \
0     1000.0   3.0  00001   1.0  296.6     0.0     0.0      20.382000   
1     1000.0   3.0  00001   1.0    0.0    24.0    24.0      73.148000   
2     1000.0   3.0  00001   1.0    0.0   216.0   216.0      19.764000   
3     1000.0   3.0  00001   1.0  288.0   504.0   504.0       3.219900   
5     1000.0   3.0  00001   2.0    0.0   696.0   192.0      23.210000   
...      ...   ...    ...   ...    ...     ...     ...            ...   
5371  3000.0   3.0  00200  15.0    0.0  7416.0   360.0       6.091200   
5372  3000.0   3.0  00200  15.0    0.0  7584.0  7584.0       2.043100   
5373  3000.0   3.0  00200  17.0  259.2  8064.0     0.0       0.090115   
5374  3000.0   3.0  00200  17.0    0.0  8088.0    24.0      53.990000   
5375  3000.0   3.0  00200  17.0    0.0  8256.0   192.0      17.255000   

             ID  PK_round1  
0     100000001     20.382  
1     100000001     73.148  
2     100000001     19.764  
3     100000001      0.000  
5     100000001      0.000  
...         ...        ...  
5371  300000200      0.000  
5372  300000200      0.000  
5373  300000200      0.000  
5374  300000200      0.000  
5375  300000200      0.000  

[5137 rows x 10 columns]

# Splitting Data

Now, we want to be able to split our data into training, validation, and test groups. This is a method of ensuring that we can hypertune our model without any biases and have the most robust version of our model in the final output.

We begin by defining a function called ```data_split``` that we will use eventually to split our data in a customized way. Arguments for the function include a dataframe (```df```), a column (```on_col```), a list of columns to be saved in the output dataframes (```save_cols```), the seed of the split (```seed```), and the proportion of the data that will go into the test set (```test_size```).

In [14]:
#Importing the necessary library for splitting the data into training and test sets
from sklearn.model_selection import train_test_split

#Defining the function called data_split which takes in a dataframe, a column name (likely with patient identifiers), 
# and a default test size of 20%. 
def data_split(df, on_col, save_cols=None, seed=2020, test_size=0.2):
    
    #Setting the default save_cols to all columns of the dataset
    if not save_cols:
        save_cols = df.columns.values

    #Setting a variable called target which contains all the unique values of a certain column. 
    #This variable will be helpful to split the unique patients into train and test later in the code.
    target = df[on_col].unique()
    
    #Setting the train variable to contain the random unique patients numbers of 80% of the patients 
    #while the remaining 20% is the allocated to the test variable
    train, test = train_test_split(target, random_state=seed, test_size=test_size, shuffle=True)
    
    #Creating the train and test dataframes (train_df and test_df). The training dataframe (train_df) includes all rows of 
    #patients in the train variable (train). The test dataframe (test_df) includes all rows of remaining patients (test).
    train_df = df[df[on_col].isin(train)]
    test_df = df[df[on_col].isin(test)]

    #Returning  
    return train_df[save_cols], test_df[save_cols]

We then build the code for the ```main``` that takes in our input data uses the helper function that we created above ```data_split()``` to split our data into training, validation, and test sets.

Next, the main adds a specific portion of the test set into the training and validation sets that we created above. 

Then, the main performs an augmentation that splits each patient into three groups. The main purpose of this augmentation in my opinion is to eliminate the problems associated with translating the results from one dosing regimen to the next.


In [15]:
def datasplitter(data, fold, model):
  train, test = data_split(data, "PTNM", seed=1329+fold, test_size=0.2)
  train, validate = data_split(train, "PTNM", seed=1329+fold+model, test_size=0.2)

  test_add_to_train = pd.DataFrame()
  test_add_to_train = pd.concat([test_add_to_train, test[(test.DSFQ == 1) & (test.TIME < 168)]], ignore_index=True)
  test_add_to_train = pd.concat([test_add_to_train, test[(test.DSFQ == 3) & (test.TIME < 504)]], ignore_index=True)
  train = pd.concat([train, test_add_to_train], ignore_index=True)
  validate = pd.concat([validate, test_add_to_train], ignore_index=True)

  # James' augmentation
  # Deep learning is prone to overfitting, and they applied augmentation to prevent overfitting. We applied 
  # timewise truncation to increase the number of training examples. For each training example, in addition to the 
  # original example, we also truncated the examples at 1008 hr, 1512 hr, and 2016 hr and generated and added a 
  # set of new examples to the training examples.

  augment_data = pd.DataFrame(columns=train.columns)

  for ptnm in train.PTNM.unique():
      df = train[(train.PTNM == ptnm) & (train.TIME <= 2 * 21 * 24) & (train.TIME >= 0)]
      df["PTNM"] = df["PTNM"] + str(0.1)
      augment_data = pd.concat([augment_data, df], ignore_index=True)

      df = train[(train.PTNM == ptnm) & (train.TIME <= 3 * 21 * 24) & (train.TIME >= 0)]
      df["PTNM"] = df["PTNM"] + str(0.2)
      augment_data = pd.concat([augment_data, df], ignore_index=True)

      df = train[(train.PTNM == ptnm) & (train.TIME <= 4 * 21 * 24) & (train.TIME >= 0)]
      df["PTNM"] = df["PTNM"] + str(0.3)
      augment_data = pd.concat([augment_data, df], ignore_index=True)

  train = pd.concat([train, augment_data], ignore_index=True).reset_index(drop=True)
  

  train.to_csv("train.csv", index=False) 
  validate.to_csv("validate.csv", index=False)
  test.to_csv("test.csv", index=False)
  return train, validate, test

In [13]:
# for fold in np.arange(1,6):
#   for model in np.arange(1,6):
#      train, validate, test = datasplitter(data_complete, fold, model)
#     #  runtest()
#      print(fold, model)

#ENDDDDDDD


# BRINGING IN THE ENTIRE UTILS.PY FILE

In [16]:
!pip install torchdiffeq
import os
import logging
import torch
import torch.nn as nn
import numpy as np
from sklearn.metrics import mean_squared_error, r2_score
from torchdiffeq import odeint_adjoint as odeint


def makedirs(dirname):
    if not os.path.exists(dirname):
        os.makedirs(dirname)


def get_device(tensor):
    device = torch.device("cuda")
    if tensor.is_cuda:
        device = tensor.get_device()
    return device


def load_model(ckpt_path, encoder=None, ode_func=None, classifier=None, device="cuda"):
    if not os.path.exists(ckpt_path):
        raise Exception("Checkpoint " + ckpt_path + " does not exist.")

    checkpt = torch.load(ckpt_path)
    if encoder is not None:
        encoder_state = checkpt["encoder"]
        encoder.load_state_dict(encoder_state)
        encoder.to(device)

    if ode_func is not None:
        ode_state = checkpt["ode"]
        ode_func.load_state_dict(ode_state)
        ode_func.to(device)

    if classifier is not None:
        classifier_state = checkpt["classifier"]
        classifier.load_state_dict(classifier_state)
        classifier.to(device)


def get_logger(logpath, filepath, package_files=[], displaying=True, saving=True, debug=False):
    logger = logging.getLogger()
    if debug:
        level = logging.DEBUG
    else:
        level = logging.INFO
    logger.setLevel(level)
    if saving:
        info_file_handler = logging.FileHandler(logpath, mode="a")
        info_file_handler.setLevel(level)
        logger.addHandler(info_file_handler)
    if displaying:
        console_handler = logging.StreamHandler()
        console_handler.setLevel(level)
        logger.addHandler(console_handler)
    logger.info(filepath)
    """
    with open(filepath, "r") as f:
        logger.info(f.read())

    for f in package_files:
        logger.info(f)
        with open(f, "r") as package_f:
            logger.info(package_f.read())
    """

    return logger


def inf_generator(iterable):
    """Allows training with DataLoaders in a single infinite loop:
        for i, (x, y) in enumerate(inf_generator(train_loader)):
    """
    iterator = iterable.__iter__()
    while True:
        try:
            yield iterator.__next__()
        except StopIteration:
            iterator = iterable.__iter__()


def init_network_weights(net, std = 0.1):
    for m in net.modules():
        if isinstance(m, nn.Linear):
            nn.init.normal_(m.weight, mean=0, std=std)
            nn.init.constant_(m.bias, val=0)


def reverse(tensor):
    idx = [i for i in range(tensor.size(0)-1, -1, -1)]
    return tensor[idx]


def sample_standard_gaussian(mu, sigma):
    device = get_device(mu)
    d = torch.distributions.normal.Normal(
            torch.Tensor([0.]).to(device),
            torch.Tensor([1.]).to(device))
    r = d.sample(mu.size()).squeeze(-1)
    return r * sigma.float() + mu.float()


def compute_loss_on_train(criterion, labels, preds):
    preds = preds.permute(1, 0, 2)

    idx_not_nan = ~(torch.isnan(labels).to(device) | (labels == -1))
    # print(idx_not_nan)
    preds = preds[idx_not_nan]
    labels = labels[idx_not_nan]

    return torch.sqrt(criterion(preds, labels)).to(device)


def compute_loss_on_test(encoder, ode_func, classifier, dataloader, n_batches, device, phase):
    ptnms = []
    Times = torch.Tensor([]).to(device=device)
    predictions = torch.Tensor([]).to(device=device)
    ground_truth = torch.Tensor([]).to(device=device)
    latent_dim = 6

    for itr in range(n_batches):
        ptnm, times, features, labels, cmax_time = dataloader.__next__()
        dosing = torch.zeros([features.size(0), features.size(1), latent_dim]).to(device)
        dosing[:, :, 0] = features[:, :, -2]
        dosing = dosing.permute(1, 0, 2).to(device)

        encoder_out = encoder(features)
        qz0_mean, qz0_var = encoder_out[:, :latent_dim], encoder_out[:, latent_dim:]
        z0 = sample_standard_gaussian(qz0_mean, qz0_var)

        solves = z0.unsqueeze(0).clone()
        try:
            for idx, (time0, time1) in enumerate(zip(times[:-1], times[1:])):
                z0 += dosing[idx]
                time_interval = torch.Tensor([time0 - time0, time1 - time0]).to(device)
                sol = odeint(ode_func, z0, time_interval, rtol=1e-4, atol=1e-4)
                z0 = sol[-1].clone()
                solves = torch.cat([solves, sol[-1:, :]], 0).to(device)
        except AssertionError:
            print(times)
            print(time0, time1, time_interval, ptnm)
            continue
    
        preds = classifier(solves, cmax_time).permute(1, 0, 2)


        if phase == "test":
            idx_not_nan = ~(torch.isnan(labels).to(device) | (labels == -1))
            preds = preds[idx_not_nan]
            labels = labels[idx_not_nan]


            times = times[idx_not_nan.flatten()]
            ptnms += ptnm * len(times)
            Times = torch.cat((Times, times*24)).to(device)

            predictions = torch.cat((predictions, preds)).to(device)
            ground_truth = torch.cat((ground_truth, labels)).to(device)
        
        else:
            """
            time_idx = (times >= 21)
            preds = preds[:, time_idx, :]
            labels = labels[:, time_idx, :]
            """
            idx_not_nan = ~(torch.isnan(labels).to(device) | (labels == -1))
            preds = preds[idx_not_nan]
            labels = labels[idx_not_nan]

            predictions = torch.cat((predictions, preds))
            ground_truth = torch.cat((ground_truth, labels))

    rmse_loss = mean_squared_error(
        ground_truth.cpu().numpy(), predictions.cpu().numpy(),
        squared=False
    )
    r2 = r2_score(ground_truth.cpu().numpy(), predictions.cpu().numpy())

    if phase == "test":
        return {"PTNM": ptnms,
                "TIME": Times.cpu(),  
                "labels": ground_truth.cpu().tolist(), 
                "preds": predictions.cpu().tolist(),
                "loss": rmse_loss}
    else:
        return {"labels": ground_truth.cpu().tolist(), 
                "preds": predictions.cpu().tolist(),
                "loss": rmse_loss,
                "r2": r2}


def compute_loss_on_interp(encoder, ode_func, classifier, dataloader, dataloader_o, n_batches, device, phase):
    ptnms = []
    Times = torch.Tensor([]).to(device=device)
    predictions = torch.Tensor([]).to(device=device)
    ground_truth = torch.Tensor([]).to(device=device)
    latent_dim = 6

    for itr in range(n_batches):
        ptnm, times, features, labels, cmax_time = dataloader.__next__()
        ptnm_o, times_o, features_o, labels_o, cmax_time_o = dataloader_o.__next__()
        assert ptnm == ptnm_o

        dosing = torch.zeros([features.size(0), features.size(1), latent_dim]).to(device)
        dosing[:, :, 0] = features[:, :, -2]
        dosing = dosing.permute(1, 0, 2)

        encoder_out = encoder(features_o)
        qz0_mean, qz0_var = encoder_out[:, :latent_dim], encoder_out[:, latent_dim:]
        z0 = sample_standard_gaussian(qz0_mean, qz0_var)

        solves = z0.unsqueeze(0).clone()
        try:
            for idx, (time0, time1) in enumerate(zip(times[:-1], times[1:])):
                z0 += dosing[idx]
                time_interval = torch.Tensor([time0 - time0, time1 - time0]).to(device)
                sol = odeint(ode_func, z0, time_interval, rtol=1e-4, atol=1e-4)
                z0 = sol[-1].clone()
                solves = torch.cat([solves, sol[-1:, :]], 0).to(device)
        except AssertionError:
            print(times)
            print(time0, time1, time_interval, ptnm)
            continue
    
        preds = classifier(solves, cmax_time).permute(1, 0, 2)

        idx_not_nan = ~(torch.isnan(labels).to(device) | (labels == -1))
        print(idx_not_nan)
        preds = preds[idx_not_nan]
        labels = labels[idx_not_nan]

        times = times[idx_not_nan.flatten()]
        ptnms += ptnm * len(times)
        Times = torch.cat((Times, times*24)).to(device)

        predictions = torch.cat((predictions, preds)).to(device)
        ground_truth = torch.cat((ground_truth, labels)).to(device)
        
    rmse_loss = mean_squared_error(
        ground_truth.cpu().numpy(), predictions.cpu().numpy(),
        squared=False
    )
    r2 = r2_score(ground_truth.cpu().numpy(), predictions.cpu().numpy())

    return {"PTNM": ptnms,
            "TIME": Times,  
            "labels": ground_truth.cpu().tolist(), 
            "preds": predictions.cpu().tolist(),
            "loss": rmse_loss}


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


# BRINGING IN THE ENTIRE MODEL.PY FILE

In [15]:
import torch
import torch.nn as nn
# from args import args
# from torch.nn.modules.rnn import GRU, LSTM, RNN
# import utils


class ODEFunc(nn.Module):

    def __init__(self, input_dim, hidden_dim):
        super(ODEFunc, self).__init__()

        self.net = nn.Sequential(
            nn.Linear(input_dim, hidden_dim),
            nn.SELU(),
            nn.Linear(hidden_dim, hidden_dim),
            nn.SELU(),
            nn.Linear(hidden_dim, hidden_dim),
            nn.SELU(),
            nn.Linear(hidden_dim, input_dim)
        )

        for m in self.net.modules():
            if isinstance(m, nn.Linear):
                nn.init.normal_(m.weight, mean=0, std=0.001)
                nn.init.constant_(m.bias, val=0.5)

    def forward(self, t, x):
        # print(x)
        return self.net(x)


class Encoder(nn.Module):

    def __init__(self, input_dim, output_dim, hidden_dim, device=torch.device("cuda")):
        super(Encoder, self).__init__()

        self.output_dim = output_dim
        self.input_dim = input_dim
        self.hidden_dim = hidden_dim
        self.device = device

        self.hiddens_to_output = nn.Sequential(
            nn.Linear(self.hidden_dim, self.hidden_dim),
            nn.ReLU(),
            nn.Linear(self.hidden_dim, self.output_dim),
        )
        init_network_weights(self.hiddens_to_output, std=0.001)

        # self.rnn = nn.RNN(self.input_dim, self.hidden_dim, nonlinearity="relu").to(device)
        self.rnn = nn.GRU(self.input_dim, self.hidden_dim).to(device)

    def forward(self, data):
        data = data.permute(1, 0, 2)
        data = reverse(data)
        output_rnn, _ = self.rnn(data)
        #print(output_rnn)
        outputs = self.hiddens_to_output(output_rnn[-1])
        #print(outputs)
        
        return outputs


class Classifier(nn.Module):

    def __init__(self, latent_dim, output_dim):
        super(Classifier, self).__init__()
        self.net = nn.Sequential(
            nn.Linear(latent_dim + 20, 32),
            nn.SELU(),
            nn.Linear(32, output_dim),
        )
        
        init_network_weights(self.net, std=0.001)

    def forward(self, z, cmax_time):
        cmax_time = cmax_time.repeat(z.size(0), 1, 1)
        z = torch.cat([z, cmax_time], 2)
        return self.net(z)


# BRINGING IN THE ENTIRE EVALUATION.PY FILE

In [16]:
import pandas as pd
from sklearn.metrics import r2_score, mean_squared_error
from scipy.stats import pearsonr


def score(df, label_col, pred_col, score_fn):
    y_true = df[label_col].values
    y_pred = df[pred_col].values
    if score_fn is mean_squared_error:
        return score_fn(y_true, y_pred, squared=False)
    elif score_fn is pearsonr:
        return score_fn(y_true, y_pred)[0]
    else:
        return score_fn(y_true, y_pred)
    

def merge_predictions(files, method="mean"):
    cols = ['PTNM','TIME','preds']
    left = pd.read_csv(files[0])
    for f in files[1:]:
        right = pd.read_csv(f)
        left = left.merge(right[cols], on=["PTNM", "TIME"], how="left")
    preds = [col for col in left.columns.values if col.startswith("preds")]
    left["pred_agg"] = left[preds].agg(method, axis=1)

    ref = pd.read_csv("data.csv")[["PTNM", "DSFQ"]].drop_duplicates() 
    # print(ref)
    left = left.merge(ref, on=["PTNM"], how="left")
    print(left.shape)
    left_q1w = left[(left.DSFQ == 1) & (left.TIME >= 168)]
    left_q3w = left[(left.DSFQ == 3) & (left.TIME >= 504)]
    return pd.concat([left_q1w, left_q3w], ignore_index=False)


def write_score(pred_df, fold, model_name, rmse_dict, score_fn):
    res = score(pred_df, "labels", "pred_agg", score_fn)
    if not rmse_dict.get("fold", []) or rmse_dict["fold"][-1] != fold:
        rmse_dict["fold"] = rmse_dict.get("fold", []) + [fold]
    rmse_dict[model_name] = rmse_dict.get(model_name, []) + [round(res, 5)]


def main(score_type, score_fn, folds, models):
    records = {}
    for fold in folds:
        in_file = ["fold_{}/fold_{}_model_{}.csv".format(fold, fold, m) for m in models]
        predictions = merge_predictions(in_file)
        predictions.to_csv("predictions.csv", index=False)
        write_score(predictions, fold, "ensemble (mean)", records, score_fn)

        for f in in_file:
            predictions = merge_predictions([f])
            write_score(predictions, fold, f.split("/")[1][7:], records, score_fn)
    
    df = pd.DataFrame(records)
    print(df)
    summary_df = df.drop(columns="fold").agg(["min", "max", "mean", "median"])
    print(summary_df)
    df.to_csv(f"{score_type}.txt", sep="\t", index=False)
    summary_df.to_csv(f"{score_type}_summary.txt", sep="\t", index=False)


  # main("rmse", mean_squared_error, args)
  # main("r2", r2_score, args)
  # main("correlation", pearsonr, args)


# BRINGING IN THE ENTIRE DATA_PARSE.PY FILE

In [17]:
import torch
from torch.utils.data import Dataset, DataLoader
import numpy as np
import pandas as pd
import random
from sklearn.model_selection import train_test_split

class TDM1(Dataset):

    def __init__(self, data_to_load, label_col, feature_cols, device, phase="train"):
        self.data = pd.read_csv(data_to_load)

        self.label_col = label_col
        self.features = feature_cols
        self.device = device
        self.phase = phase
        self.data["TIME"] = self.data["TIME"] / 24

    def __len__(self):
        return self.data.PTNM.unique().shape[0] 

    def __getitem__(self, index):
        ptnm = self.data.PTNM.unique()[index]
        cur_data = self.data[self.data["PTNM"] == ptnm]
        times, features, labels, cmax_time = self.process(cur_data)
        return ptnm, times, features, labels, cmax_time

    def process(self, data):
        data = data.reset_index(drop=True)
        """
        if self.phase == "train":
            random_time = np.random.randint(low=21, high=int(data["TIME"].max()) + 1)
            data = data[data.TIME <= random_time]
        else:
            pass
        """
        if (data.DSFQ == 1).all():
            cmax_time = data.loc[data.TIME < 7, ["TIME", "PK_timeCourse"]].values.flatten()
            cmax = data.loc[data.TIME < 7, "PK_timeCourse"].max()
        else:
            cmax_time = data.loc[data.TIME < 21, ["TIME", "PK_timeCourse"]].values.flatten()
            cmax = data.loc[data.TIME < 21, "PK_timeCourse"].max()

        cmax_time_full = np.zeros((20, ))
        if len(cmax_time) <= 20:
            cmax_time_full[:len(cmax_time)] = cmax_time
        else:
            cmax_time_full[:] = cmax_time[:20]
        
        data["PK_round1"] = data["PK_round1"] / cmax

        features = data[self.features].values
        labels = data[self.label_col].values
        times = data["TIME"].values

        times = torch.from_numpy(times)
        features = torch.from_numpy(features)
        labels = torch.from_numpy(labels).unsqueeze_(-1)
        cmax_time_full = torch.from_numpy(cmax_time_full)
        return times, features, labels, cmax_time_full


def tdm1_collate_fn(batch, device):
    D = batch[0][2].shape[1]
    N = 1

    combined_tt, inverse_indices = torch.unique(torch.cat([ex[1] for ex in batch]),
        sorted=True, return_inverse=True)
    combined_tt = combined_tt.to(device)
    # print(combined_tt, inverse_indices)

    offset = 0
    combined_features = torch.zeros([len(batch), len(combined_tt), D]).to(device)
    combined_label = torch.zeros([len(batch), len(combined_tt), N]).to(device)
    combined_label[:] = np.nan
    combined_cmax_time = torch.zeros([len(batch), 20]).to(device)
    # print(combined_label.shape)

    ptnms = []
    for b, (ptnm, tt, features, label, cmax_time) in enumerate(batch):
        ptnms.append(ptnm)
        tt = tt.to(device)
        features = features.to(device)
        label = label.to(device)
        cmax_time = cmax_time.to(device)

        indices = inverse_indices[offset:offset + len(tt)]
        offset += len(tt)

        combined_features[b, indices] = features.float()
        combined_label[b, indices] = label.float()
        combined_cmax_time[b, :] = cmax_time.float()
    combined_tt = combined_tt.float()

    return ptnms, combined_tt, combined_features, combined_label, combined_cmax_time


def parse_tdm1(device, phase="train"):
    train_data_path = "train.csv"
    val_data_path = "validate.csv"
    test_data_path = "test.csv"

    feature_cols = ['TFDS','TIME','CYCL','AMT',"PK_round1"]
    """
    covariates = ['SEX','AGE','WT','RACR','RACE','BSA',
                  'BMI','ALBU','TPRO','WBC','CRCL','CRET',
                  'SGOT','SGPT','TBIL','TMBD','ALKP', 'HER', 
                  'ECOG','KEOALL','ASIAN']
    feature_cols += covariates
    """
    label_col = "PK_timeCourse"
    train = TDM1(train_data_path, label_col, feature_cols, device, phase="train")
    validate = TDM1(val_data_path, label_col, feature_cols, device, phase="validate")
    test = TDM1(test_data_path, label_col, feature_cols, device, phase="test")

    ptnm, times, features, labels, cmax_time = train[0]
    input_dim = features.size(-1)
    # n_labels = 1

    if phase == "train":
        train_dataloader = DataLoader(train, batch_size=1, shuffle=True, 
            collate_fn=lambda batch: tdm1_collate_fn(batch, device))
        val_dataloader = DataLoader(validate, batch_size=1, shuffle=False,
            collate_fn=lambda batch: tdm1_collate_fn(batch, device))

        dataset_objs = {
            "train_dataloader": inf_generator(train_dataloader),
            "val_dataloader": inf_generator(val_dataloader),
            "n_train_batches": len(train_dataloader),
            "n_val_batches": len(val_dataloader),
            "input_dim": input_dim
        }

    else:
        test_dataloader = DataLoader(test, batch_size=1, shuffle=False,
            collate_fn=lambda batch: tdm1_collate_fn(batch, device))

        dataset_objs = {
            "test_dataloader": inf_generator(test_dataloader),
            "n_test_batches": len(test_dataloader),
            "input_dim": input_dim
        }

    return dataset_objs

    




# if __name__ == "__main__":
#     print("run")
#     data = parse_tdm1("cpu")
#     for ptnms, times, features, labels, cmax_time in data["train_dataloader"]:
#         print(ptnms)
#         print(times)
#         print(features)
#         print(labels)
#         print(cmax_time)
#         break


# BRINGING IN THE ENTIRE RUNTRAIN.PY FILE

In [18]:
import os
import sys
import numpy as np
import pandas as pd
from random import SystemRandom
from tqdm import tqdm

import torch
import torch.nn as nn
import torch.optim as optim
from torchdiffeq import odeint_adjoint as odeint
from scipy.stats import pearsonr

# import utils
# from model import *
# from data_parse import parse_tdm1
# from args import args


device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
file_name = os.path.basename("__file__")[:-3]


input_cmd = sys.argv
input_cmd = " ".join(input_cmd)

########################################################################
## Main runnings
def runtraining(random_seed, model, fold, lr=0.00005, l2=0.1, epochs=30, tol=1e-4, continue_train=False):
  save = f"Fold{fold}"
  makedirs(save)
  torch.manual_seed(random_seed + model + fold)
  np.random.seed(random_seed + model + fold)

  ckpt_path = os.path.join(save, f"fold_{fold}_model_{model}.ckpt")
  ########################################################################
  tdm1_obj = parse_tdm1(device, phase="train")
  input_dim = tdm1_obj["input_dim"]
  hidden_dim = 128
  latent_dim = 6

  encoder = Encoder(input_dim=input_dim, output_dim=2 * latent_dim, hidden_dim=hidden_dim).to(device)
  ode_func = ODEFunc(input_dim=latent_dim, hidden_dim=16).to(device)
  classifier = Classifier(latent_dim=latent_dim, output_dim=1).to(device)

  torch.save({'encoder': encoder.state_dict(), 'ode': ode_func.state_dict(), 'classifier': classifier.state_dict()}, ckpt_path)

  if continue_train:
      load_model(ckpt_path, encoder, ode_func, classifier, device)

  ########################################################################
  ## Train
  log_path = "logs/" + f"fold_{fold}_model_{model}.log"
  makedirs("logs/")
  logger = get_logger(logpath=log_path, filepath=os.path.abspath("__file__"))
  logger.info(input_cmd)

  batches_per_epoch = tdm1_obj["n_train_batches"]
  criterion = nn.MSELoss().to(device=device)
  params = (list(encoder.parameters()) + 
            list(ode_func.parameters()) + 
            list(classifier.parameters()))
  optimizer = optim.Adam(params, lr=lr, weight_decay=l2)
  best_rmse = 0x7fffffff
  best_epochs = 0

  for epoch in range(1, epochs):

      for _ in tqdm(range(batches_per_epoch), ascii=True):
          optimizer.zero_grad()

          ptnms, times, features, labels, cmax_time = tdm1_obj["train_dataloader"].__next__()
          dosing = torch.zeros([features.size(0), features.size(1), latent_dim]).to(device)
          dosing[:, :, 0] = features[:, :, -2]
          dosing = dosing.permute(1, 0, 2)

          encoder_out = encoder(features)
          qz0_mean, qz0_var = encoder_out[:, :latent_dim], encoder_out[:, latent_dim:]
          z0 = sample_standard_gaussian(qz0_mean, qz0_var).to(device)
          
          solves = z0.unsqueeze(0).clone()
          try:
              for idx, (time0, time1) in enumerate(zip(times[:-1], times[1:])):
                  z0 += dosing[idx]
                  # print(z0.device)
                  time_interval = torch.Tensor([time0 - time0, time1 - time0])
                  sol = odeint(ode_func, z0, time_interval, rtol=tol, atol=tol)
                  z0 = sol[-1].clone()
                  solves = torch.cat([solves, sol[-1:, :]], 0)
          except AssertionError:
              print(times)
              print(time0, time1, time_interval, ptnms)
              continue
      
          preds = classifier(solves, cmax_time)

          loss = compute_loss_on_train(criterion, labels, preds)
          try: 
              loss.backward()
          except RuntimeError:
              print(ptnms)
              print(times)
              continue
          optimizer.step()
      
      idx_not_nan = ~(torch.isnan(labels) | (labels == -1))
      preds = preds.permute(1, 0, 2)[idx_not_nan]
      labels = labels[idx_not_nan]
      print(preds)
      print(labels)

      with torch.no_grad():
          
          train_res = compute_loss_on_test(encoder, ode_func, classifier,
              tdm1_obj["train_dataloader"], tdm1_obj["n_train_batches"], 
              device, phase="train")

          validation_res = compute_loss_on_test(encoder, ode_func, classifier,
              tdm1_obj["val_dataloader"], tdm1_obj["n_val_batches"], 
              device, phase="validate")
          
          train_loss = train_res["loss"] 
          validation_loss = validation_res["loss"]
          if validation_loss < best_rmse:
              torch.save({'encoder': encoder.state_dict(),
                          'ode': ode_func.state_dict(),
                          'classifier': classifier.state_dict()}, ckpt_path)
              best_rmse = validation_loss
              best_epochs = epoch

          message = """
          Epoch {:04d} | Training loss {:.6f} | Training R2 {:.6f} | Validation loss {:.6f} | Validation R2 {:.6f}
          Best loss {:.6f} | Best epoch {:04d}
          """.format(epoch, train_loss, train_res["r2"], validation_loss, validation_res["r2"], best_rmse, best_epochs)
          logger.info(message)
            

#BRINGING IN THE ENTIRE RUNPREDICT.PY FILE

In [19]:
import os
import numpy as np
import pandas as pd
import torch



device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

###############################################################
## Main runnings
def runpredict(random_seed, model, fold, lr=0.00005, l2=0.1, epochs=30, tol=1e-4, continue_train=True):
  save = f"Fold{fold}"
  makedirs(save)
  ckpt_path = os.path.join(save, f"fold_{fold}_model_{model}.ckpt")
  eval_path = os.path.join(save, f"fold_{fold}_model_{model}.csv")
  res_path = "rmse.csv"

  ########################################################################
  tdm1_obj = parse_tdm1(device, phase="test")
  input_dim = tdm1_obj["input_dim"]
  hidden_dim = 128 
  latent_dim = 6

  encoder = Encoder(input_dim=input_dim, output_dim=2 * latent_dim, hidden_dim=hidden_dim)
  ode_func = ODEFunc(input_dim=latent_dim, hidden_dim=16)
  classifier = Classifier(latent_dim=latent_dim, output_dim=1)

  load_model(ckpt_path, encoder, ode_func, classifier, device)

  ########################################################################
  ## Predict & Evaluate
  with torch.no_grad():
      test_res = compute_loss_on_test(encoder, ode_func, classifier,
          tdm1_obj["test_dataloader"], tdm1_obj["n_test_batches"], 
          device, phase="test")
      

  eval_results = pd.DataFrame(test_res).drop(columns="loss")
  eval_results.to_csv(eval_path, index=False)

  """
  with torch.no_grad():
      test_res = utils.compute_loss_on_interp(encoder, ode_func, classifier, args,
          tdm1_obj["interp_dataloader"], tdm1_obj["test_dataloader"], tdm1_obj["n_interp_batches"], 
          device, phase="test")

  eval_results = pd.DataFrame(test_res).drop(columns="loss")
  eval_results.to_csv(eval_path + ".interp", index=False)

  with torch.no_grad():
      test_res = utils.compute_loss_on_interp(encoder, ode_func, classifier, args,
          tdm1_obj["nodosing_dataloader"], tdm1_obj["test_dataloader"], tdm1_obj["n_interp_batches"], 
          device, phase="test")

  eval_results = pd.DataFrame(test_res).drop(columns="loss")
  eval_results.to_csv(eval_path + ".nodosing", index=False)
  """

## Light GBM Model

In [18]:
fold, model = 1,1
train, validate, test = datasplitter(data_complete, fold, model)
x_train = train[["CYCL", "AMT", "TIME", "TFDS"]]
y_train = train["PK_timeCourse"]
x_val = validate[["CYCL", "AMT", "TIME", "TFDS"]]
y_val = validate["PK_timeCourse"]
x_test = test[["CYCL", "AMT", "TIME", "TFDS"]]
y_test = test["PK_timeCourse"]
train

<ipython-input-15-3e15d65d0aaa>:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["PTNM"] = df["PTNM"] + str(0.1)
<ipython-input-15-3e15d65d0aaa>:25: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["PTNM"] = df["PTNM"] + str(0.2)
<ipython-input-15-3e15d65d0aaa>:29: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide

STUD  DSFQ      PTNM  CYCL    AMT   TIME   TFDS  PK_timeCourse  \
0     1000.0   3.0     00001   1.0  296.6    0.0    0.0        20.3820   
1     1000.0   3.0     00001   1.0    0.0   24.0   24.0        73.1480   
2     1000.0   3.0     00001   1.0    0.0  216.0  216.0        19.7640   
3     1000.0   3.0     00001   1.0  288.0  504.0  504.0         3.2199   
4     1000.0   3.0     00001   2.0    0.0  696.0  192.0        23.2100   
...      ...   ...       ...   ...    ...    ...    ...            ...   
8062  3000.0   3.0  001960.2   1.0    0.0  360.0  360.0         5.2152   
8063  3000.0   3.0  001960.3   1.0  217.0    0.0    0.0         1.7780   
8064  3000.0   3.0  001960.3   1.0    0.0   24.0   24.0        53.9210   
8065  3000.0   3.0  001960.3   1.0    0.0  192.0  192.0        16.2500   
8066  3000.0   3.0  001960.3   1.0    0.0  360.0  360.0         5.2152   

             ID  PK_round1  
0     100000001    20.3820  
1     100000001    73.1480  
2     100000001    19.7640  
3     100000001     0.0000  
4     100000001     0.0000  
...         ...        ...  
8062  300000196     5.2152  
8063  300000196     1.7780  
8064  300000196    53.9210  
8065  300000196    16.2500  
8066  300000196     5.2152  

[8067 rows x 10 columns]

In [ ]:
import lightgbm as lgb
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
from scipy.stats import pearsonr

# Assuming you have trained the model and made predictions on the validation set


params = {
    'boosting_type': 'gbdt',
    'learning_rate': 0.005,
    'num_leaves': 30,
    'num_trees': 1000,
    'objective': 'regression',
    'metric': 'rmse'
}  

model = lgb.LGBMRegressor(
    boosting_type= 'gbdt',
    learning_rate=0.005,
    num_leaves= 30,
    num_trees=1000,
    objective='regression',
    num_boost_round=10000,
    early_stopping_rounds=50, 
    verbose_eval=100)

model.fit(x_train,y_train,eval_set=[(x_test,y_test),(x_train,y_train)],
          verbose=20,eval_metric='rmse')

# train_data = lgb.Dataset(train_x, train_y)

# # Train the model
# model = lgb.train(train_data,boosting_type= 'gbdt',
#     learning_rate=0.005,
#     num_leaves= 30,
#     num_trees=1000,
#     objective='regression',
#     metric='rmse',num_boost_round=10000, 
#     valid_sets=[train_data, (val_x, val_y)], 
#                   early_stopping_rounds=50, 
#                   verbose_eval=100)

y_pred = model.predict(x_test)

# Evaluate the model using RMSE
rmse = mean_squared_error(y_test, y_pred, squared=False)
print(f"RMSE: {rmse:.4f}")

y_pred = model.predict(x_test)
r2 = r2_score(y_test, y_pred)
print("R2 Score: {:.2f}".format(r2))

pearson_corr, _ = pearsonr(y_test, y_pred)
print(f"Pearson correlation coefficient: {pearson_corr:.4f}")

/usr/local/lib/python3.8/dist-packages/lightgbm/engine.py:118: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))
/usr/local/lib/python3.8/dist-packages/lightgbm/engine.py:123: UserWarning: Found `early_stopping_rounds` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))


Training until validation scores don't improve for 50 rounds.
[20]	training's l2: 328.546	training's rmse: 18.1258	valid_0's l2: 323.106	valid_0's rmse: 17.9751
[40]	training's l2: 282.82	training's rmse: 16.8173	valid_0's l2: 276.832	valid_0's rmse: 16.6383
[60]	training's l2: 245.4	training's rmse: 15.6653	valid_0's l2: 239.024	valid_0's rmse: 15.4604
[80]	training's l2: 214.705	training's rmse: 14.6528	valid_0's l2: 208.116	valid_0's rmse: 14.4262
[100]	training's l2: 189.398	training's rmse: 13.7622	valid_0's l2: 182.609	valid_0's rmse: 13.5133
[120]	training's l2: 168.651	training's rmse: 12.9866	valid_0's l2: 162.021	valid_0's rmse: 12.7287
[140]	training's l2: 151.538	training's rmse: 12.3101	valid_0's l2: 144.759	valid_0's rmse: 12.0316
[160]	training's l2: 137.474	training's rmse: 11.7249	valid_0's l2: 130.743	valid_0's rmse: 11.4343
[180]	training's l2: 125.898	training's rmse: 11.2204	valid_0's l2: 119.009	valid_0's rmse: 10.9091
[200]	training's l2: 116.355	training's rmse:

In [ ]:
# for fold in np.arange(1,6):
#   for model in np.arange(1,6):
#      train, validate, test = datasplitter(data_complete, fold, model)
#      runtraining(1000, model, fold, 0.00005, 0.1, 30, 1e-4, True)
#      runpredict(1000, model, fold, 0.00005, 0.1, 30, 1e-4, True)

#      print(fold, model)

## LSTM Model

In [80]:
lstm_train_x = train[["CYCL", "AMT", "TIME", "TFDS", "PK_round1"]]
lstm_train_y = train[["PK_timeCourse"]]
twentyelem = torch.flatten(torch.tensor(lstm_train_x[["TIME", "PK_round1"]].head(20).values, dtype=torch.float32))

CYCL    AMT   TIME   TFDS  PK_round1
0      1.0  296.6    0.0    0.0    20.3820
1      1.0    0.0   24.0   24.0    73.1480
2      1.0    0.0  216.0  216.0    19.7640
3      1.0  288.0  504.0  504.0     0.0000
4      2.0    0.0  696.0  192.0     0.0000
...    ...    ...    ...    ...        ...
8062   1.0    0.0  360.0  360.0     5.2152
8063   1.0  217.0    0.0    0.0     1.7780
8064   1.0    0.0   24.0   24.0    53.9210
8065   1.0    0.0  192.0  192.0    16.2500
8066   1.0    0.0  360.0  360.0     5.2152

[8067 rows x 5 columns]

In [81]:
class LSTM(nn.Module):
      def __init__(self, input_dim, hidden_dim, output_dim):
        super(LSTM, self).__init__()
        
        self.hidden_dim = hidden_dim
        
        # LSTM layers 1 and 2
        self.lstm1 = nn.LSTM(input_dim, hidden_dim, batch_first=True)
        self.lstm2 = nn.LSTM(hidden_dim, hidden_dim, batch_first=True)
        
        # Decoder
        self.decodernet = nn.Sequential(
            nn.Linear(hidden_dim, 128),
            nn.ReLU(), 
            nn.Linear(128, output_dim))
        init_network_weights(self.decodernet)
      
      def forward(self, x):
        output, _ = self.lstm1(x)
        output = torch.tanh(output)
        output, _ = self.lstm2(output)
        output = torch.tanh(output)

        #concatenate with PK_cycle1
        print(output)
        #output = torch.cat((output, twentyelem))

        output = self.decodernet(output)

        return output


# Define hyperparameters
input_dim = 5
hidden_dim = 128
output_dim = 1
batch_size = 1
learning_rate = 0.00005
weight_decay = 0.1
num_epochs = 30
seq_len = 5

#initializing everything
model = LSTM(input_dim, hidden_dim, output_dim)
optimizer = torch.optim.Adam(model.parameters(), lr=0.00005, weight_decay=0.1)
criterion = nn.MSELoss()


#train loop
for epoch in range(num_epochs):
  optimizer.zero_grad()
  x = torch.tensor(lstm_train_x.values, dtype=torch.float32)
  y = torch.tensor(lstm_train_y.values, dtype=torch.float32)

  output = model(x)
  loss = criterion(output, y)
  loss.backward()
  optimizer.step()
        
  print('Epoch [{}/{}], Batch [{}/{}], Loss: {:.4f}'.format(epoch+1, num_epochs, epoch+1, len(train), loss.item()**(0.5)))

tensor([[ 0.0017, -0.0051,  0.0344,  ..., -0.0531,  0.0190, -0.0219],
        [-0.0264, -0.0072,  0.0043,  ..., -0.0111,  0.0648,  0.0107],
        [ 0.0279,  0.0375,  0.0151,  ..., -0.0276,  0.1076,  0.0025],
        ...,
        [ 0.0093,  0.0076, -0.0041,  ..., -0.0462,  0.1677,  0.0554],
        [ 0.0459,  0.0171,  0.0054,  ..., -0.0587,  0.1835,  0.0393],
        [ 0.0633,  0.0309,  0.0131,  ..., -0.0643,  0.1954,  0.0293]],
       grad_fn=<TanhBackward0>)
Epoch [1/30], Batch [1/8067], Loss: 27.3611
tensor([[ 0.0018, -0.0055,  0.0341,  ..., -0.0529,  0.0193, -0.0222],
        [-0.0263, -0.0076,  0.0038,  ..., -0.0110,  0.0653,  0.0104],
        [ 0.0283,  0.0370,  0.0145,  ..., -0.0274,  0.1081,  0.0022],
        ...,
        [ 0.0095,  0.0072, -0.0051,  ..., -0.0460,  0.1685,  0.0552],
        [ 0.0463,  0.0167,  0.0044,  ..., -0.0584,  0.1842,  0.0390],
        [ 0.0640,  0.0306,  0.0123,  ..., -0.0639,  0.1961,  0.0290]],
       grad_fn=<TanhBackward0>)
Epoch [2/30], Batch [2/8

RUNNING THE NEURAL-ODE MODEL

In [ ]:
fold, model = 1,1
train, validate, test = datasplitter(data_complete, fold, model)
runtraining(1000, model, fold, 0.00005, 0.1, 30, 1e-4, False)
runpredict(1000, model, fold, 0.00005, 0.1, 30, 1e-4, True)

print(fold, model)

<ipython-input-33-3e15d65d0aaa>:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["PTNM"] = df["PTNM"] + str(0.1)
<ipython-input-33-3e15d65d0aaa>:25: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["PTNM"] = df["PTNM"] + str(0.2)
<ipython-input-33-3e15d65d0aaa>:29: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide

tensor([ 0.2262,  5.2315,  8.9786,  9.3791, 13.0919, 13.3213, 13.5220, 17.2932,
        17.4656], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([ 5.8156,  0.8816, 16.1310,  0.8871, 19.8420,  3.7824,  0.8870, 20.1590,
         5.8156], device='cuda:0')


INFO:root:
          Epoch 0001 | Training loss 26.023926 | Training R2 -0.761777 | Validation loss 27.330231 | Validation R2 -1.235342
          Best loss 27.330231 | Best epoch 0001
          

          Epoch 0001 | Training loss 26.023926 | Training R2 -0.761777 | Validation loss 27.330231 | Validation R2 -1.235342
          Best loss 27.330231 | Best epoch 0001
          

          Epoch 0001 | Training loss 26.023926 | Training R2 -0.761777 | Validation loss 27.330231 | Validation R2 -1.235342
          Best loss 27.330231 | Best epoch 0001
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [05:25<00:00,  2.06it/s]


tensor([0.9692, 4.4988, 4.6919, 7.1922, 7.3790], device='cuda:0',
       grad_fn=<IndexBackward0>)
tensor([0.1803, 6.8826, 0.1879, 5.1060, 0.1803], device='cuda:0')


INFO:root:
          Epoch 0002 | Training loss 26.541620 | Training R2 -0.832568 | Validation loss 28.543949 | Validation R2 -1.438291
          Best loss 27.330231 | Best epoch 0001
          

          Epoch 0002 | Training loss 26.541620 | Training R2 -0.832568 | Validation loss 28.543949 | Validation R2 -1.438291
          Best loss 27.330231 | Best epoch 0001
          

          Epoch 0002 | Training loss 26.541620 | Training R2 -0.832568 | Validation loss 28.543949 | Validation R2 -1.438291
          Best loss 27.330231 | Best epoch 0001
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [05:25<00:00,  2.06it/s]


tensor([1.2331, 1.9829, 1.9415, 1.8990, 1.8555, 1.6702, 1.5474, 1.4219, 1.3320,
        1.2420, 1.8636, 1.6413, 1.5813, 1.4314, 1.2515, 1.8606, 1.6507, 1.4408,
        1.2609], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([0.1027, 4.3675, 3.3207, 2.6789, 2.2198, 1.1160, 0.6726, 0.3426, 0.2066,
        0.1245, 4.4427, 1.1442, 0.8164, 0.3513, 0.1277, 4.4458, 1.1451, 0.3516,
        0.1278], device='cuda:0')


INFO:root:
          Epoch 0003 | Training loss 26.121161 | Training R2 -0.774967 | Validation loss 27.962708 | Validation R2 -1.340000
          Best loss 27.330231 | Best epoch 0001
          

          Epoch 0003 | Training loss 26.121161 | Training R2 -0.774967 | Validation loss 27.962708 | Validation R2 -1.340000
          Best loss 27.330231 | Best epoch 0001
          

          Epoch 0003 | Training loss 26.121161 | Training R2 -0.774967 | Validation loss 27.962708 | Validation R2 -1.340000
          Best loss 27.330231 | Best epoch 0001
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [05:25<00:00,  2.07it/s]


tensor([ 4.4198,  9.1870,  9.0119,  8.8579, 12.9042, 12.7374, 12.5940],
       device='cuda:0', grad_fn=<IndexBackward0>)
tensor([ 1.3031, 17.1530,  4.2189,  1.2679, 17.6280,  4.3359,  1.3031],
       device='cuda:0')


INFO:root:
          Epoch 0004 | Training loss 23.783239 | Training R2 -0.471456 | Validation loss 24.067657 | Validation R2 -0.733505
          Best loss 24.067657 | Best epoch 0004
          

          Epoch 0004 | Training loss 23.783239 | Training R2 -0.471456 | Validation loss 24.067657 | Validation R2 -0.733505
          Best loss 24.067657 | Best epoch 0004
          

          Epoch 0004 | Training loss 23.783239 | Training R2 -0.471456 | Validation loss 24.067657 | Validation R2 -0.733505
          Best loss 24.067657 | Best epoch 0004
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [05:27<00:00,  2.05it/s]


tensor([13.5571, 18.6401, 18.2470, 17.3657, 22.4104, 21.7328, 19.8837],
       device='cuda:0', grad_fn=<IndexBackward0>)
tensor([17.6690, 69.9310, 19.6240,  2.0229, 70.4890, 19.8330,  2.0444],
       device='cuda:0')


INFO:root:
          Epoch 0005 | Training loss 20.085930 | Training R2 -0.049516 | Validation loss 21.196657 | Validation R2 -0.344597
          Best loss 21.196657 | Best epoch 0005
          

          Epoch 0005 | Training loss 20.085930 | Training R2 -0.049516 | Validation loss 21.196657 | Validation R2 -0.344597
          Best loss 21.196657 | Best epoch 0005
          

          Epoch 0005 | Training loss 20.085930 | Training R2 -0.049516 | Validation loss 21.196657 | Validation R2 -0.344597
          Best loss 21.196657 | Best epoch 0005
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [05:26<00:00,  2.06it/s]


tensor([18.2134, 27.3814, 26.7597], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([19.2920, 85.6310, 19.2920], device='cuda:0')


INFO:root:
          Epoch 0006 | Training loss 22.338833 | Training R2 -0.298154 | Validation loss 27.210217 | Validation R2 -1.215753
          Best loss 21.196657 | Best epoch 0005
          

          Epoch 0006 | Training loss 22.338833 | Training R2 -0.298154 | Validation loss 27.210217 | Validation R2 -1.215753
          Best loss 21.196657 | Best epoch 0005
          

          Epoch 0006 | Training loss 22.338833 | Training R2 -0.298154 | Validation loss 27.210217 | Validation R2 -1.215753
          Best loss 21.196657 | Best epoch 0005
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [05:25<00:00,  2.06it/s]


tensor([4.2663, 8.3554, 7.1731, 6.1392, 9.9788, 6.3810, 9.9749, 6.0479, 9.5008,
        5.2280], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([ 2.5158, 10.0050,  2.5792,  0.8070, 40.4740,  0.8047, 40.4720,  0.8047,
        39.6060,  0.7875], device='cuda:0')


INFO:root:
          Epoch 0007 | Training loss 22.732279 | Training R2 -0.344285 | Validation loss 28.738533 | Validation R2 -1.471648
          Best loss 21.196657 | Best epoch 0005
          

          Epoch 0007 | Training loss 22.732279 | Training R2 -0.344285 | Validation loss 28.738533 | Validation R2 -1.471648
          Best loss 21.196657 | Best epoch 0005
          

          Epoch 0007 | Training loss 22.732279 | Training R2 -0.344285 | Validation loss 28.738533 | Validation R2 -1.471648
          Best loss 21.196657 | Best epoch 0005
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [05:27<00:00,  2.05it/s]


tensor([ 4.5063,  8.0869,  7.9028, 12.6300, 10.3681,  9.9557, 14.9023, 11.9170,
        11.4565, 12.9553, 12.4464, 17.1199, 12.9465, 17.8168, 15.3982],
       device='cuda:0', grad_fn=<IndexBackward0>)
tensor([ 8.4665,  2.6608,  2.2691, 18.5760,  3.2225,  2.3435, 25.5770,  3.2261,
         2.3462,  3.2262,  2.3463, 25.6730,  2.3549, 30.2890,  8.4665],
       device='cuda:0')


INFO:root:
          Epoch 0008 | Training loss 20.586393 | Training R2 -0.102468 | Validation loss 25.082453 | Validation R2 -0.882771
          Best loss 21.196657 | Best epoch 0005
          

          Epoch 0008 | Training loss 20.586393 | Training R2 -0.102468 | Validation loss 25.082453 | Validation R2 -0.882771
          Best loss 21.196657 | Best epoch 0005
          

          Epoch 0008 | Training loss 20.586393 | Training R2 -0.102468 | Validation loss 25.082453 | Validation R2 -0.882771
          Best loss 21.196657 | Best epoch 0005
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [05:25<00:00,  2.06it/s]


tensor([18.4772, 30.6279, 29.3546], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([ 7.2340, 79.9040, 30.3950], device='cuda:0')


INFO:root:
          Epoch 0009 | Training loss 19.352571 | Training R2 0.025722 | Validation loss 22.767998 | Validation R2 -0.551340
          Best loss 21.196657 | Best epoch 0005
          

          Epoch 0009 | Training loss 19.352571 | Training R2 0.025722 | Validation loss 22.767998 | Validation R2 -0.551340
          Best loss 21.196657 | Best epoch 0005
          

          Epoch 0009 | Training loss 19.352571 | Training R2 0.025722 | Validation loss 22.767998 | Validation R2 -0.551340
          Best loss 21.196657 | Best epoch 0005
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [05:24<00:00,  2.07it/s]


tensor([10.6994, 16.9112, 14.9006, 12.8969, 11.1788, 11.3401, 16.5176, 14.0705,
        11.2851, 15.6539, 13.5866, 11.2876, 10.9350, 14.6185, 12.5573, 10.3170,
         9.9756, 13.2317, 11.1341,  8.8862,  8.5468, 11.5985,  9.4430,  7.1543,
         6.8109,  9.6637,  7.4329,  4.7200,  7.4382,  5.1189,  2.2956,  4.7944,
         2.3672, -0.6183,  2.1842,  0.0199], device='cuda:0',
       grad_fn=<IndexBackward0>)
tensor([ 7.2829, 42.8440, 12.3400,  4.1313,  1.6174,  1.6724, 20.9970,  5.9940,
         1.7166, 17.7690,  5.9424,  1.9898,  1.7019, 17.6790,  5.9126,  1.9798,
         1.6933, 17.6670,  5.9084,  1.9784,  1.6921, 17.9310,  5.9966,  2.0079,
         1.7174, 17.9400,  5.9999,  1.7183, 17.9410,  6.0000,  1.7184, 17.4120,
         5.8233,  1.6677, 18.6260,  7.2829], device='cuda:0')


INFO:root:
          Epoch 0010 | Training loss 19.532093 | Training R2 0.007563 | Validation loss 24.344080 | Validation R2 -0.773553
          Best loss 21.196657 | Best epoch 0005
          

          Epoch 0010 | Training loss 19.532093 | Training R2 0.007563 | Validation loss 24.344080 | Validation R2 -0.773553
          Best loss 21.196657 | Best epoch 0005
          

          Epoch 0010 | Training loss 19.532093 | Training R2 0.007563 | Validation loss 24.344080 | Validation R2 -0.773553
          Best loss 21.196657 | Best epoch 0005
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [05:25<00:00,  2.07it/s]


tensor([17.3204, 27.0179, 26.7337, 26.4454, 26.1525, 24.9293, 23.9463, 22.5232,
        21.4830], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([12.0510, 55.0980, 40.4280, 31.6660, 25.7020, 12.4280,  7.3585,  3.6646,
         2.1727], device='cuda:0')


INFO:root:
          Epoch 0011 | Training loss 18.453506 | Training R2 0.114144 | Validation loss 21.703226 | Validation R2 -0.409633
          Best loss 21.196657 | Best epoch 0005
          

          Epoch 0011 | Training loss 18.453506 | Training R2 0.114144 | Validation loss 21.703226 | Validation R2 -0.409633
          Best loss 21.196657 | Best epoch 0005
          

          Epoch 0011 | Training loss 18.453506 | Training R2 0.114144 | Validation loss 21.703226 | Validation R2 -0.409633
          Best loss 21.196657 | Best epoch 0005
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [05:26<00:00,  2.06it/s]


tensor([18.5868, 28.4867, 28.2081, 27.9261, 27.6406, 26.7597, 33.5045, 32.9966,
        31.9769, 37.9906, 37.2009, 36.4114, 40.3970, 46.0221, 44.5810, 49.8841,
        48.3809, 53.2554, 52.0044, 57.2274, 55.9244, 61.1085, 59.7537, 64.4007,
        62.9986, 67.6103, 66.1610, 70.7374, 69.2412], device='cuda:0',
       grad_fn=<IndexBackward0>)
tensor([20.7140, 46.6930, 36.3750, 29.6610, 24.7630, 15.1220, 59.5840, 39.0430,
        20.0970, 63.5550, 35.3180, 21.6870, 21.4910, 52.0110, 22.1980, 50.6250,
        21.6590, 41.1430, 21.2610, 42.8590, 22.1220, 43.3930, 22.4050, 41.0620,
        21.2290, 40.3330, 20.8420, 40.0930, 20.7140], device='cuda:0')


INFO:root:
          Epoch 0012 | Training loss 17.937012 | Training R2 0.163038 | Validation loss 19.241276 | Validation R2 -0.107963
          Best loss 19.241276 | Best epoch 0012
          

          Epoch 0012 | Training loss 17.937012 | Training R2 0.163038 | Validation loss 19.241276 | Validation R2 -0.107963
          Best loss 19.241276 | Best epoch 0012
          

          Epoch 0012 | Training loss 17.937012 | Training R2 0.163038 | Validation loss 19.241276 | Validation R2 -0.107963
          Best loss 19.241276 | Best epoch 0012
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [05:25<00:00,  2.07it/s]


tensor([ 8.3696, 15.8541, 11.7758, 18.5145, 16.5960, 12.9726, 18.6058, 17.4685,
        13.4790, 17.8538, 13.7082, 17.7735, 13.7695, 13.4620, 17.4751, 12.9937,
        16.9394, 12.6193, 12.2878, 16.1473, 11.3268, 15.1017, 10.4692, 10.1137,
         8.5905, 12.1794,  6.8640,  4.7324,  8.0993,  5.2807,  2.4694,  5.9670,
         3.0664], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([ 9.4905, 23.3070,  3.3688, 62.3120, 21.5050,  3.5695, 43.4420, 24.6750,
         3.5667, 25.0220,  3.6169, 24.3010,  4.0329,  3.5126, 24.4200,  3.5299,
        24.5050,  4.0669,  3.5422, 24.5080,  3.5425, 24.5080,  4.0673,  3.5426,
         3.5090, 24.4200,  3.5298,  3.4397, 24.1540,  9.1820,  3.4914, 24.9650,
         9.4905], device='cuda:0')


INFO:root:
          Epoch 0013 | Training loss 17.951038 | Training R2 0.161729 | Validation loss 21.723827 | Validation R2 -0.412310
          Best loss 19.241276 | Best epoch 0012
          

          Epoch 0013 | Training loss 17.951038 | Training R2 0.161729 | Validation loss 21.723827 | Validation R2 -0.412310
          Best loss 19.241276 | Best epoch 0012
          

          Epoch 0013 | Training loss 17.951038 | Training R2 0.161729 | Validation loss 21.723827 | Validation R2 -0.412310
          Best loss 19.241276 | Best epoch 0012
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [05:24<00:00,  2.07it/s]


tensor([ 3.9080, 12.0202,  7.7294, 15.1353, 13.1146,  9.5496, 16.0494, 14.2740,
        11.9010, 10.4081], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([ 0.9887, 13.2440,  1.0495, 44.2920, 11.1530,  1.0742, 43.9160, 13.4380,
         2.8234,  1.0649], device='cuda:0')


INFO:root:
          Epoch 0014 | Training loss 17.661663 | Training R2 0.188537 | Validation loss 20.252714 | Validation R2 -0.227507
          Best loss 19.241276 | Best epoch 0012
          

          Epoch 0014 | Training loss 17.661663 | Training R2 0.188537 | Validation loss 20.252714 | Validation R2 -0.227507
          Best loss 19.241276 | Best epoch 0012
          

          Epoch 0014 | Training loss 17.661663 | Training R2 0.188537 | Validation loss 20.252714 | Validation R2 -0.227507
          Best loss 19.241276 | Best epoch 0012
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [05:25<00:00,  2.07it/s]


tensor([16.4841, 29.9286, 23.8466, 20.3671, 28.0211], device='cuda:0',
       grad_fn=<IndexBackward0>)
tensor([16.1690, 59.8340,  5.3050,  1.4892, 16.1690], device='cuda:0')


INFO:root:
          Epoch 0015 | Training loss 17.730440 | Training R2 0.182205 | Validation loss 21.766275 | Validation R2 -0.417835
          Best loss 19.241276 | Best epoch 0012
          

          Epoch 0015 | Training loss 17.730440 | Training R2 0.182205 | Validation loss 21.766275 | Validation R2 -0.417835
          Best loss 19.241276 | Best epoch 0012
          

          Epoch 0015 | Training loss 17.730440 | Training R2 0.182205 | Validation loss 21.766275 | Validation R2 -0.417835
          Best loss 19.241276 | Best epoch 0012
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [05:25<00:00,  2.07it/s]


tensor([31.8013, 35.0596, 34.8570, 34.6486, 33.9801, 36.5707, 35.5430, 34.6229,
        36.0843, 35.1592, 37.4872, 35.6183, 37.9322, 36.0413, 38.3187, 36.4047,
        38.6757, 36.7381, 39.0361, 37.0741, 39.3674, 37.3811, 39.2297],
       device='cuda:0', grad_fn=<IndexBackward0>)
tensor([66.9750, 34.8300, 30.4850, 27.1220, 19.3810, 59.4210, 39.4690, 28.2220,
        45.4640, 32.5130, 71.1690, 34.2430, 72.7160, 35.0360, 72.9860, 35.1970,
        73.1260, 35.2700, 73.7120, 35.5420, 73.9550, 35.6660, 66.9750],
       device='cuda:0')


INFO:root:
          Epoch 0016 | Training loss 17.640852 | Training R2 0.190449 | Validation loss 21.638426 | Validation R2 -0.401228
          Best loss 19.241276 | Best epoch 0012
          

          Epoch 0016 | Training loss 17.640852 | Training R2 0.190449 | Validation loss 21.638426 | Validation R2 -0.401228
          Best loss 19.241276 | Best epoch 0012
          

          Epoch 0016 | Training loss 17.640852 | Training R2 0.190449 | Validation loss 21.638426 | Validation R2 -0.401228
          Best loss 19.241276 | Best epoch 0012
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [05:26<00:00,  2.06it/s]


tensor([21.9314, 41.9099, 38.6898, 30.8984, 43.4602, 40.9532, 36.0212, 44.2726,
        41.7171, 36.9887], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([18.0000, 80.3400, 25.8200,  4.2027, 84.0030, 27.2700,  4.4388, 82.8660,
        26.9150,  4.3809], device='cuda:0')


INFO:root:
          Epoch 0017 | Training loss 18.278654 | Training R2 0.130852 | Validation loss 24.355211 | Validation R2 -0.775175
          Best loss 19.241276 | Best epoch 0012
          

          Epoch 0017 | Training loss 18.278654 | Training R2 0.130852 | Validation loss 24.355211 | Validation R2 -0.775175
          Best loss 19.241276 | Best epoch 0012
          

          Epoch 0017 | Training loss 18.278654 | Training R2 0.130852 | Validation loss 24.355211 | Validation R2 -0.775175
          Best loss 19.241276 | Best epoch 0012
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [05:26<00:00,  2.06it/s]


tensor([10.3748, 18.5630, 15.3926, 11.8126, 20.5219, 17.9761, 16.8433, 15.2561,
        12.5712], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([17.2550, 16.6050,  5.5696,  2.1836, 53.8670, 17.2450, 10.7980,  5.7840,
         2.2677], device='cuda:0')


INFO:root:
          Epoch 0018 | Training loss 17.109701 | Training R2 0.238465 | Validation loss 19.479559 | Validation R2 -0.135575
          Best loss 19.241276 | Best epoch 0012
          

          Epoch 0018 | Training loss 17.109701 | Training R2 0.238465 | Validation loss 19.479559 | Validation R2 -0.135575
          Best loss 19.241276 | Best epoch 0012
          

          Epoch 0018 | Training loss 17.109701 | Training R2 0.238465 | Validation loss 19.479559 | Validation R2 -0.135575
          Best loss 19.241276 | Best epoch 0012
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [05:47<00:00,  1.93it/s]


tensor([ 2.1460, 19.7407, 14.7389, 12.0465, 10.4475, 21.7934, 19.3598, 16.0997,
        15.2626, 23.6322, 20.3157, 18.6914, 16.6881, 23.7784, 21.2274, 17.9135],
       device='cuda:0', grad_fn=<IndexBackward0>)
tensor([ 0.7684, 22.7900,  2.1049,  0.8118,  0.5042, 22.8300,  5.4673,  0.8133,
         0.5051, 28.9870,  4.3086,  1.6618,  0.5051, 21.5700,  5.1656,  0.7684],
       device='cuda:0')


INFO:root:
          Epoch 0019 | Training loss 16.719358 | Training R2 0.272816 | Validation loss 18.523756 | Validation R2 -0.026870
          Best loss 18.523756 | Best epoch 0019
          

          Epoch 0019 | Training loss 16.719358 | Training R2 0.272816 | Validation loss 18.523756 | Validation R2 -0.026870
          Best loss 18.523756 | Best epoch 0019
          

          Epoch 0019 | Training loss 16.719358 | Training R2 0.272816 | Validation loss 18.523756 | Validation R2 -0.026870
          Best loss 18.523756 | Best epoch 0019
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [06:10<00:00,  1.82it/s]


tensor([12.6779, 35.9059, 29.9472, 17.9949, 33.2749, 21.1599], device='cuda:0',
       grad_fn=<IndexBackward0>)
tensor([ 1.5969, 68.9360, 20.1160,  1.3031, 20.2650,  1.3127], device='cuda:0')


INFO:root:
          Epoch 0020 | Training loss 16.239904 | Training R2 0.313924 | Validation loss 18.098284 | Validation R2 0.019761
          Best loss 18.098284 | Best epoch 0020
          

          Epoch 0020 | Training loss 16.239904 | Training R2 0.313924 | Validation loss 18.098284 | Validation R2 0.019761
          Best loss 18.098284 | Best epoch 0020
          

          Epoch 0020 | Training loss 16.239904 | Training R2 0.313924 | Validation loss 18.098284 | Validation R2 0.019761
          Best loss 18.098284 | Best epoch 0020
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [06:40<00:00,  1.68it/s]


tensor([11.1487, 35.6937, 25.9895, 11.8030, 24.6132, 10.6339], device='cuda:0',
       grad_fn=<IndexBackward0>)
tensor([ 2.0389, 55.6560, 16.1950,  2.1895, 15.9180,  2.1521], device='cuda:0')


INFO:root:
          Epoch 0021 | Training loss 14.238553 | Training R2 0.472604 | Validation loss 14.747548 | Validation R2 0.349126
          Best loss 14.747548 | Best epoch 0021
          

          Epoch 0021 | Training loss 14.238553 | Training R2 0.472604 | Validation loss 14.747548 | Validation R2 0.349126
          Best loss 14.747548 | Best epoch 0021
          

          Epoch 0021 | Training loss 14.238553 | Training R2 0.472604 | Validation loss 14.747548 | Validation R2 0.349126
          Best loss 14.747548 | Best epoch 0021
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [07:42<00:00,  1.45it/s]


tensor([ 6.0025, 21.2748,  9.7366,  2.9221, -1.2914, -1.8394,  6.8513, -0.5222,
        -3.7404, -4.1694,  4.2793, -1.5292, -5.0112, -5.3974,  3.2380, -2.9071,
        -5.7227, -6.0990,  2.5686, -2.9882, -6.7219,  0.0590, -3.5582, -7.2824,
         1.4858, -4.5743, -7.3604, -7.7334, -6.2319, -8.3242,  6.5506, -1.4723,
        -6.2639, -8.4843], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([ 0.4168, 38.0910,  7.1683,  1.7787,  0.4414,  0.3499, 11.5080,  1.7944,
         0.4453,  0.3530, 10.8300,  2.1303,  0.4190,  0.3322, 10.7820,  1.6812,
         0.4172,  0.3307, 10.7730,  2.1190,  0.3304,  6.7693,  2.1190,  0.3304,
        10.7810,  1.6811,  0.4171,  0.3307,  1.0556,  0.3304, 35.8960,  6.7693,
         1.3316,  0.4168], device='cuda:0')


INFO:root:
          Epoch 0022 | Training loss 12.984377 | Training R2 0.561421 | Validation loss 14.248682 | Validation R2 0.392415
          Best loss 14.248682 | Best epoch 0022
          

          Epoch 0022 | Training loss 12.984377 | Training R2 0.561421 | Validation loss 14.248682 | Validation R2 0.392415
          Best loss 14.248682 | Best epoch 0022
          

          Epoch 0022 | Training loss 12.984377 | Training R2 0.561421 | Validation loss 14.248682 | Validation R2 0.392415
          Best loss 14.248682 | Best epoch 0022
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [09:02<00:00,  1.24it/s]


tensor([18.0998, 40.6213, 24.4990, 12.9450, 35.1566, 21.8113, 12.9836, 35.5856,
        22.3656, 13.7323, 36.4429, 23.1427], device='cuda:0',
       grad_fn=<IndexBackward0>)
tensor([17.8830, 55.3190, 16.9870,  2.1271, 56.8470, 17.5230,  2.1942, 57.7130,
        17.7900,  2.2277, 58.0120, 17.8830], device='cuda:0')


INFO:root:
          Epoch 0023 | Training loss 11.494071 | Training R2 0.656321 | Validation loss 10.316972 | Validation R2 0.681462
          Best loss 10.316972 | Best epoch 0023
          

          Epoch 0023 | Training loss 11.494071 | Training R2 0.656321 | Validation loss 10.316972 | Validation R2 0.681462
          Best loss 10.316972 | Best epoch 0023
          

          Epoch 0023 | Training loss 11.494071 | Training R2 0.656321 | Validation loss 10.316972 | Validation R2 0.681462
          Best loss 10.316972 | Best epoch 0023
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [09:29<00:00,  1.18it/s]


tensor([10.0309, 22.6771,  4.3416, 35.5102, 17.2923,  4.0489, 31.9729, 19.4330,
         3.9611, 19.5964,  3.9186], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([ 9.4905, 23.3070,  3.3688, 62.3120, 21.5050,  3.5695, 43.4420, 24.6750,
         3.5667, 25.0220,  3.6169], device='cuda:0')


INFO:root:
          Epoch 0024 | Training loss 9.125366 | Training R2 0.783376 | Validation loss 8.817200 | Validation R2 0.767342
          Best loss 8.817200 | Best epoch 0024
          

          Epoch 0024 | Training loss 9.125366 | Training R2 0.783376 | Validation loss 8.817200 | Validation R2 0.767342
          Best loss 8.817200 | Best epoch 0024
          

          Epoch 0024 | Training loss 9.125366 | Training R2 0.783376 | Validation loss 8.817200 | Validation R2 0.767342
          Best loss 8.817200 | Best epoch 0024
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [09:08<00:00,  1.23it/s]


tensor([11.9240, 42.9376, 17.1037,  6.2021, 36.8456, 14.8831,  6.2175, 15.1140,
         6.2897, 36.2234, 14.6992,  6.3002], device='cuda:0',
       grad_fn=<IndexBackward0>)
tensor([ 6.8146, 52.2980, 17.1410,  3.0146, 53.6910, 17.8200,  3.1340, 18.0930,
         3.1820, 52.6960, 17.4970,  3.0772], device='cuda:0')


INFO:root:
          Epoch 0025 | Training loss 8.054012 | Training R2 0.831255 | Validation loss 7.481621 | Validation R2 0.832487
          Best loss 7.481621 | Best epoch 0025
          

          Epoch 0025 | Training loss 8.054012 | Training R2 0.831255 | Validation loss 7.481621 | Validation R2 0.832487
          Best loss 7.481621 | Best epoch 0025
          

          Epoch 0025 | Training loss 8.054012 | Training R2 0.831255 | Validation loss 7.481621 | Validation R2 0.832487
          Best loss 7.481621 | Best epoch 0025
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [09:42<00:00,  1.15it/s]


tensor([28.2936, 41.2426, 37.7484, 34.9178, 28.4797, 45.4059, 34.9197, 28.4063,
        34.9162, 28.2926, 44.6815, 28.0385, 44.4297, 27.7856, 43.9892, 27.4481,
        43.6655, 26.9595, 43.4320, 26.7604, 43.2279, 26.6278, 40.0453],
       device='cuda:0', grad_fn=<IndexBackward0>)
tensor([66.9750, 34.8300, 30.4850, 27.1220, 19.3810, 59.4210, 39.4690, 28.2220,
        45.4640, 32.5130, 71.1690, 34.2430, 72.7160, 35.0360, 72.9860, 35.1970,
        73.1260, 35.2700, 73.7120, 35.5420, 73.9550, 35.6660, 66.9750],
       device='cuda:0')


INFO:root:
          Epoch 0026 | Training loss 8.236652 | Training R2 0.823515 | Validation loss 9.226169 | Validation R2 0.745258
          Best loss 7.481621 | Best epoch 0025
          

          Epoch 0026 | Training loss 8.236652 | Training R2 0.823515 | Validation loss 9.226169 | Validation R2 0.745258
          Best loss 7.481621 | Best epoch 0025
          

          Epoch 0026 | Training loss 8.236652 | Training R2 0.823515 | Validation loss 9.226169 | Validation R2 0.745258
          Best loss 7.481621 | Best epoch 0025
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [10:05<00:00,  1.11it/s]


tensor([ 6.7199, 58.4880, 49.9393, 42.3200, 17.7389,  9.8955,  4.2216,  2.1865,
        54.5666, 20.1374,  5.4704,  2.8662, 20.7267,  6.4120,  3.5570, 21.6857,
         7.6034,  4.4656, 22.0683,  9.3576,  5.3775, 22.8746,  9.1226,  5.9426,
         9.8813], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([ 4.2077, 59.4320, 46.9050, 38.1280, 14.2580,  7.9220,  3.6186,  1.1171,
        59.4890, 17.3990,  4.4157,  1.1207, 17.1950,  4.3640,  1.1075, 17.2810,
         4.3856,  1.1130, 16.8630,  5.2057,  1.0861, 16.8560,  4.2779,  1.0857,
         4.2077], device='cuda:0')


INFO:root:
          Epoch 0027 | Training loss 7.497412 | Training R2 0.853773 | Validation loss 6.695976 | Validation R2 0.865821
          Best loss 6.695976 | Best epoch 0027
          

          Epoch 0027 | Training loss 7.497412 | Training R2 0.853773 | Validation loss 6.695976 | Validation R2 0.865821
          Best loss 6.695976 | Best epoch 0027
          

          Epoch 0027 | Training loss 7.497412 | Training R2 0.853773 | Validation loss 6.695976 | Validation R2 0.865821
          Best loss 6.695976 | Best epoch 0027
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [10:36<00:00,  1.06it/s]


tensor([ 3.6727, 40.1708, 34.6549, 28.0243, 22.8536, 11.4611, 46.3199, 27.0008,
        14.0570, 49.7140, 28.3959, 14.9416, 50.8246, 41.8635, 15.0637, 41.9609,
        15.1325, 42.0220, 15.1861, 15.1296, 50.1058, 15.1032, 41.7337, 15.0848],
       device='cuda:0', grad_fn=<IndexBackward0>)
tensor([12.5700, 31.8440, 25.2520, 20.7650, 17.3880, 10.5840, 40.6760, 22.8730,
        13.9690, 43.5540, 24.6480, 15.0630, 44.0460, 35.6960, 15.2700, 35.8460,
        15.3380, 35.8950, 15.3590, 15.2690, 43.9280, 15.2390, 35.5910, 15.2300],
       device='cuda:0')


INFO:root:
          Epoch 0028 | Training loss 6.617049 | Training R2 0.886097 | Validation loss 6.445163 | Validation R2 0.875685
          Best loss 6.445163 | Best epoch 0028
          

          Epoch 0028 | Training loss 6.617049 | Training R2 0.886097 | Validation loss 6.445163 | Validation R2 0.875685
          Best loss 6.445163 | Best epoch 0028
          

          Epoch 0028 | Training loss 6.617049 | Training R2 0.886097 | Validation loss 6.445163 | Validation R2 0.875685
          Best loss 6.445163 | Best epoch 0028
          
  0%|          | 0/672 [00:00<?, ?it/s]/usr/local/lib/python3.8/dist-packages/torchdiffeq/_impl/misc.py:296: UserWarning: t is not on the same device as y0. Coercing to y0.device.
  warnings.warn("t is not on the same device as y0. Coercing to y0.device.")
100%|##########| 672/672 [10:27<00:00,  1.07it/s]


tensor([10.6736, 80.7203, 28.2796,  9.4597,  5.9675, 19.9985,  5.9450, 60.4802,
        23.6570,  7.9908,  5.6307, 68.2241,  5.6427, 23.0071,  7.6063,  4.9287,
        67.9367, 22.9141,  7.5226,  4.6280, 67.8351,  6.9415,  4.1676, 66.8014,
        21.8695,  6.9581,  3.8048, 64.9581, 20.9832,  6.7995,  3.6086, 55.4373,
        21.2905,  6.6753,  3.3315, 62.7740, 24.0949,  6.9778], device='cuda:0',
       grad_fn=<IndexBackward0>)
tensor([ 6.1430, 77.5480, 24.0970,  7.7805,  2.9525, 17.9790,  3.0428, 55.8410,
        21.1480,  6.8284,  3.0454, 63.6220,  2.9330, 20.3360,  6.5662,  2.9284,
        63.5080, 20.3310,  6.5647,  2.9278, 63.5290,  6.5668,  2.9287, 62.7370,
        20.0850,  6.4851,  2.8923, 61.1750, 19.5850,  6.3236,  2.8202, 51.6830,
        19.5730,  6.3199,  2.8186, 59.4280, 22.3600,  6.1430], device='cuda:0')


INFO:root:
          Epoch 0029 | Training loss 6.451946 | Training R2 0.891710 | Validation loss 6.527579 | Validation R2 0.872485
          Best loss 6.445163 | Best epoch 0028
          

          Epoch 0029 | Training loss 6.451946 | Training R2 0.891710 | Validation loss 6.527579 | Validation R2 0.872485
          Best loss 6.445163 | Best epoch 0028
          

          Epoch 0029 | Training loss 6.451946 | Training R2 0.891710 | Validation loss 6.527579 | Validation R2 0.872485
          Best loss 6.445163 | Best epoch 0028
          


1 1


In [ ]:
complete_data

NameError: ignored

In [ ]:
for fold in np.arange(1,6):
  for model in np.arange(1,6):
     train, validate, test = datasplitter(data_complete, fold, model)
     

     print(fold, model)